app/models/bookmark.rb

```diff
@@ -100,19 +100,29 @@ def clear_reminder!
     )
   end
 
+  scope :with_reminders, -> do
+    where("reminder_at IS NOT NULL")
+  end
+
   scope :pending_reminders, ->(before_time = Time.now.utc) do
-    where("reminder_at IS NOT NULL AND reminder_at <= :before_time", before_time: before_time)
+    with_reminders.where("reminder_at <= :before_time", before_time: before_time)
   end
 
   scope :pending_reminders_for_user, ->(user) do
     pending_reminders.where(user: user)
   end
 
+  scope :for_user_in_topic, ->(user_id, topic_id) {
+    joins(:post).where(user_id: user_id, posts: { topic_id: topic_id })
+  }
+
   def self.count_per_day(opts = nil)
     opts ||= {}
     result = where('bookmarks.created_at >= ?', opts[:start_date] || (opts[:since_days_ago] || 30).days.ago)
     result = result.where('bookmarks.created_at <= ?', opts[:end_date]) if opts[:end_date]
-    result = result.joins(:topic).merge(Topic.in_category_and_subcategories(opts[:category_id])) if opts[:category_id]
+    result = result.joins("INNER JOIN posts ON posts.id = bookmarks.post_id AND posts.deleted_at IS NULL")
+      .joins("INNER JOIN topics ON topics.id = posts.topic_id AND topics.deleted_at IS NULL")
+      .merge(Topic.in_category_and_subcategories(opts[:category_id])) if opts[:category_id]
```

Comment:
```suggestion
    if opts[:category_id]
      result = result
        .joins("INNER JOIN posts ON posts.id = bookmarks.post_id AND posts.deleted_at IS NULL")
        .joins("INNER JOIN topics ON topics.id = posts.topic_id AND topics.deleted_at IS NULL")
        .merge(Topic.in_category_and_subcategories(opts[:category_id]))
    end
```

The conditional at the end is a little surprising. How about the above instead? Also, I believe you can just do the following as well since we have a default scope on both `Post` and `Topic`.

```suggestion
    if opts[:category_id]
      result = result.joins(post: :topic).merge(Topic.in_category_and_subcategories(opts[:category_id]))
    end
```

Reply:
You're right on both counts, good call.